hoomd/mpcd/MPCDReversePerturbationFlow.cc

```diff
@@ -0,0 +1,339 @@
+    m_lo_pos = make_scalar2(lo_pos - delta_half, lo_pos + delta_half);
+
+    // explicit check that the two regions for velocity swapping are inside
+    // of the box entirely and don't overlap
+    if (m_hi_pos.y >= global_hi.z)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: hi slab " << m_hi_pos.y
+                                  << "is outside of box: " << global_hi.z << "." << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    if (m_lo_pos.x < global_lo.z)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: max slab " << m_lo_pos.y
+                                  << "is outside of box: " << global_lo.z << "." << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    if (m_lo_pos.y > m_hi_pos.x)
+        {
+        m_exec_conf->msg->error() << "MPCDReversePerturbationFlow: hi slab " << m_hi_pos.x
+                                  << " and lo slab " << m_lo_pos.y << " are overlapping."
+                                  << std::endl;
+        throw std::runtime_error("slab outside simulation box in MPCDReversePerturbationFlow");
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::MPCDReversePerturbationFlow::update(uint64_t timestep)
+    {
+    // reset momentum exchange for this step
+    m_momentum_exchange = Scalar(0);
+
+    // don't do anything if this is an empty group
+    const unsigned int n_mpcd = m_mpcd_pdata->getN();
+    if (n_mpcd == 0)
+        return;
+
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+std::vector<std::string> mpcd::MPCDReversePerturbationFlow::getProvidedLogQuantities()
+    {
+    std::vector<std::string> ret {"rp_momentum"};
+    return ret;
+    }
+
+/*!
+ * \param quantity Name of the log quantity to get
+ * \param timestep Current time step of the simulation
+ */
+Scalar mpcd::MPCDReversePerturbationFlow::getLogValue(const std::string& quantity,
+                                                      uint64_t timestep)
+    {
+    if (quantity == "rp_momentum")
+        {
+        return m_momentum_exchange;
+        }
+    else
+        {
+        m_exec_conf->msg->error() << "azplugins.mpcd.reverse_pertubation: " << quantity
+                                  << " is not a valid log quantity." << std::endl;
+        throw std::runtime_error("Error getting log value");
+        }
+    return Scalar(0.0);
+    }
+
+/*!
+ * Finds all particles in the "min" and "max" slab in z direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to p_target in x direction.
+ */
+void mpcd::MPCDReversePerturbationFlow::findSwapParticles()
```

Comment:
This function is the main spot where the coordinates need to be rotated, I think.